fastestimator/dataset/interleave_dataset.py

```diff
@@ -0,0 +1,271 @@
+            else:
+                random.shuffle(index_map)
+            if hasattr(ds, "fe_batch_indices"):
+                self.index_maps.append([ds.fe_batch_indices(index) for index in index_map])
+            else:
+                self.index_maps.append(index_map)
+
+    def __len__(self) -> int:
+        """Compute the length of this dataset.
+
+        Returns:
+            How many batches of data can this dataset serve per epoch. It is sum of all dataset's number of batch.
+        """
+        # first calcualte the minimum number of cycles each dataset can afford according to the repeat pattern
+        num_cycles = min(len(ds) // (f * bs) for ds, f, bs in zip(self.datasets, self.frequency, self.batch_sizes))
+        assert num_cycles > 0, "some dataset does not have enough samples for a single repeat pattern, please consider using `ExtendDataset` to increase its length"
+        # returning the sum of number of batches of each dataset
+        return sum(num_cycles * f for f in self.frequency)
+
+    def __getitem__(self, batch_idx: int) -> Dict[str, Any]:
+        """Extract items from the underlying datasets based on the given `batch_idx`.
+
+        Args:
+            batch_idx: Which the index of current batch to pull data from (or which batch_idx to query).
+
+        Raises:
+            StopIteration: when batch_idx is beyond the dataset length
+
+        Returns:
+            A list of data instance dictionaries corresponding to the current `batch_idx`. When plugged in Pipeline,
+            this dataset will return the results of operators.
+        """
+        if batch_idx >= len(self):
+            raise StopIteration
+        cumulative_cycles, cycle_pos = batch_idx // len(self.pattern), batch_idx % len(self.pattern)
+        ds_idx_now = self.pattern[cycle_pos]
+        batch_size_now, index_map_now = self.batch_sizes[ds_idx_now], self.index_maps[ds_idx_now]
+        dataset_now = self.datasets[ds_idx_now] if self.op_datasets is None else self.op_datasets[ds_idx_now]
+        sample_start_idx = cumulative_cycles * self.frequency[ds_idx_now] * batch_size_now
+        if cycle_pos > 0:
+            # if in middle of cycle, need to account for the previously used samples
+            sample_start_idx = sample_start_idx + sum(
+                [self.batch_sizes[ds_idx_now] for ds_idx in self.pattern[:cycle_pos] if ds_idx == ds_idx_now])
+        batch = []
+        for idx in range(sample_start_idx, sample_start_idx + batch_size_now):
+            item = dataset_now[index_map_now[idx]]
+            if isinstance(item, list):
+                batch.extend(item)
+            else:
+                batch.append(item)
+        return batch
+
+    def split(self,
+              *fractions: Union[float, int, Iterable[int]],
+              seed: Optional[int] = None,
+              stratify: Optional[str] = None) -> Union['InterleaveDataset', List['InterleaveDataset']]:
```

Comment:
from typing_extensions import Self

```suggestion
              stratify: Optional[str] = None) -> Union['Self', List['Self']]:
```